modules/puppet_proxy/puppet_api.rb

```diff
@@ -45,8 +45,26 @@ class Proxy::Puppet::Api < ::Sinatra::Base
       class_retriever.classes_in_environment(params[:environment]).map{|k| {k.to_s => { :name => k.name, :module => k.module, :params => k.params} } }.to_json
     rescue Proxy::Puppet::EnvironmentNotFound
       log_halt 404, "Could not find environment '#{params[:environment]}'"
+    rescue Proxy::Puppet::TimeoutError => e
+      log_halt 503, e.message
     rescue Exception => e
       log_halt 406, "Failed to show puppet classes: #{e}"
     end
   end
+
+  get "/environments/:environment/classes_and_errors" do
+    content_type :json
+    begin
+      class_retriever.classes_and_errors(params[:environment]).to_json
+    rescue NoMethodError
+      log_halt 501, "classes_and_errors api end-point is not available for '#{class_retriever.class}' provider"
+    rescue Proxy::Puppet::EnvironmentNotFound
+      log_halt 404, "Could not find environment '#{params[:environment]}'"
+    rescue Proxy::Puppet::TimeoutError => e
+      log_halt 503, e.message
+    rescue Exception => e
+      p "!!! #{e.class}"
```

Comment:
Pass `e` to `log_halt` instead (and anywhere the full exception will be useful).